build/embed.bzl

```diff
@@ -0,0 +1,267 @@
+"""Provides ability to embed assets in C++ libraries and binaries.
+
+Use the rules in this bazel files by configuring your BUILD with:
+
+  load("//build/embed.bzl", "cc_embed")
+  ...
+
+  cc_embed(
+    name = "targetname",
+    data = [
+      ":file.txt",
+      ":file.jpeg",
+      ":othertarget",
+      "//another/target",
+    ]
+  )
+
+  # Can either be cc_binary or cc_library.
+  cc_binary(
+    ...
+    deps = [
+      ":targetname",
+      ...
+    ]
+  )
+
+This will result in:
+  1) A "targetname" C++ library being created, linkable like any other library.
+  2) "file.txt", "file.jpeg", the results of building ":othertarget"
+     and "//another/target" being linked directly into this library.
+  3) A "targetname.h" file being created, defining std::string_view objects
+     containing the raw bytes of the specified files. In the example, it
+     would define: "embedded::file_txt" and "embedded::file_jpeg".
+  4) The name of the indirect target will depend on what that target is.
+     Most likely, it will contain the file exported by the target
+     (eg, "embedded::othertarget_a" if "othertarget" is a cc_library).
+
+Using embedded files is extremely easy, this is what your .cc file will look
+like:
+
+    #include "targetname.h"
+    
+    void test() {
+      std::cout << embedded::file_txt << std::endl;
+    }
+
+See the tests/ directory for more details.
+"""
+
+load("@bazel_tools//tools/cpp:toolchain_utils.bzl", "find_cpp_toolchain")
+load("@bazel_tools//tools/build_defs/cc:action_names.bzl", "CPP_LINK_STATIC_LIBRARY_ACTION_NAME")
+load("@bazel_tools//tools/build_defs/cc:action_names.bzl", "C_COMPILE_ACTION_NAME")
+
+def _clean_path(directory):
+   """Turns a bazel-out/whatever/bin/base/dir into base/dir.
+   
+   Args:
+     directory: string, the path to clean.
+
+   Returns:
+     The cleaned path, or the original path if the cleaning pattern
+     cannot be found.
+   """
+   parts = directory.split("/bin/")
+   if len(parts) > 1:
+     return _clean_string(parts[1])
+   return _clean_string(parts[0])
+
+def _clean_string(s):
+   """Replaces invalid characters with _.
+
+   This is used to turn filenames (or strings in general) into valid
+   variable names. It losely emulates the behavior of "ld".
+
+   Args:
+     s: string, path of a file to clean.
+
+   Returns:
+     The cleaned string, with invalid characters replaced.
+   """
+   # TODO: it would be better to say s/[^a-zA-Z0-9]/_/g, but starlark
+   #   does not support regexes, or python APIs that would allow this.
+   #   Recommended solution would be to build and use a tool.
+   #   Much more work. Fix it if we ever get paths with invalid characters.
+   return s.replace("/", "_").replace(".", "_").replace("-", "_").replace(" ", "_").replace("[", "_").replace("]", "_")
+
+def _cc_embed(ctx):
```

Comment:
uhm ... now that I read your text carefully, I understand the goal and mechanism, but I'm not sure about the recommended implementation.

Unless you feel strongly, I'll fix this in a followup PR.

Do you mean something like:
```
def cc_embed(ctx, **kwargs):
  # instead of using ctx.attrs.whatever
  inputs = kwargs.get("whatever", ctx.attrs.whatever)
  ...
```
or
```
def cc_embed(ctx, **kwargs):
  for attr in dir(ctx.attrs):
    kwargs.setdefault(attr, getattr(ctx.attrs, attr))
  inputs = kwargs["whatever"]
```
or 
```
def cc_embed(ctx, **kwargs):
  attrs = kwargs.get("attrs") or ctx.attrs
```
?